scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala

```diff
@@ -6,83 +6,149 @@ import io.circe.generic.auto.*
 import pureconfig.generic.derivation.default.*
 import pureconfig.ConfigReader
 import sttp.capabilities.fs2.Fs2Streams
-import uk.gov.nationalarchives.dynamoformatters.DynamoFormatters.{ArchiveFolderDynamoTable, DynamoTable, FilesTablePartitionKey, FilesTablePrimaryKey, FilesTableSortKey}
+import uk.gov.nationalarchives.dynamoformatters.DynamoFormatters.{Identifier => DynamoIdentifier, *}
+import uk.gov.nationalarchives.utils.ExternalUtils.DetailType.DR2Message
 import uk.gov.nationalarchives.ingestupsertarchivefolders.Lambda.*
 import uk.gov.nationalarchives.dp.client.Entities.{Entity, IdentifierResponse}
 import uk.gov.nationalarchives.dp.client.EntityClient.EntityType.*
 import uk.gov.nationalarchives.dp.client.EntityClient.SecurityTag.*
 import uk.gov.nationalarchives.dp.client.EntityClient.*
 import uk.gov.nationalarchives.dp.client.EntityClient
 import uk.gov.nationalarchives.dp.client.fs2.Fs2Client
-import uk.gov.nationalarchives.utils.ExternalUtils.DetailType.DR2Message
 import uk.gov.nationalarchives.utils.LambdaRunner
 import uk.gov.nationalarchives.{DADynamoDBClient, DAEventBridgeClient}
 
 import java.util.UUID
 
 class Lambda extends LambdaRunner[StepFnInput, Unit, Config, Dependencies] {
-  private val structuralObject = StructuralObject
   private val sourceId = "SourceID"
 
+  given Ordering[ArchiveFolderDynamoTable] = (x: ArchiveFolderDynamoTable, y: ArchiveFolderDynamoTable) => x.potentialParentPath.compare(y.potentialParentPath)
+
+  extension [T](potentialValue: Option[T]) private def toStringOrEmpty: String = potentialValue.map(_.toString).getOrElse("")
+
+  extension (archiveFolderDynamoTable: ArchiveFolderDynamoTable)
+    private def directParentId: Option[UUID] = archiveFolderDynamoTable.potentialParentPath.flatMap(_.split("/").lastOption).map(UUID.fromString)
+
   override def handler: (
       StepFnInput,
       Config,
       Dependencies
   ) => IO[Unit] = (stepFnInput, config, dependencies) => {
-    val logWithBatchRef = log(Map("batchRef" -> stepFnInput.batchId))(_)
 
     val folderIdPartitionKeysAndValues: List[FilesTablePrimaryKey] =
       stepFnInput.archiveHierarchyFolders.map(UUID.fromString).map(id => FilesTablePrimaryKey(FilesTablePartitionKey(id), FilesTableSortKey(stepFnInput.batchId)))
 
-    for {
-      folderRowsSortedByParentPath <- getFolderRowsSortedByParentPath(
-        dependencies.dADynamoDBClient,
-        folderIdPartitionKeysAndValues,
-        config.archiveFolderTableName
-      )
-
-      _ <- checkNumOfParentPathSlashesPerFolderIncrease(folderRowsSortedByParentPath)
-      _ <- checkEachParentPathMatchesFolderBeforeIt(folderRowsSortedByParentPath)
-
-      secretName = config.secretName
+    def getFolderRows: IO[List[ArchiveFolderDynamoTable]] =
+      dependencies.dADynamoDBClient
+        .getItems[ArchiveFolderDynamoTable, FilesTablePrimaryKey](
+          folderIdPartitionKeysAndValues,
+          config.archiveFolderTableName
+        )
+        .map(_.sorted)
```

Comment:
Ahh, I see. Should we add a test for this (to ensure that they are added in this order)? Because someone in the future could remove `.map(_.sorted)`, thinking it's not necessary.

Reply:
I tried removing the `.map(_.sorted)` and one of the tests failed so I think we're already covered.